paasd/src/service.rs

```diff
@@ -27,25 +60,64 @@ impl server_types::ProcessService for ProcessService {
     type GetLogsStream =
         Pin<Box<dyn Stream<Item = Result<LogsResponse, tonic::Status>> + Send + Sync + 'static>>;
 
-    async fn exec(&self, _req: Request<ExecRequest>) -> Result<Response<ExecResponse>, Status> {
-        Err(Status::unimplemented(""))
+    async fn exec(&self, req: Request<ExecRequest>) -> Result<Response<ExecResponse>, Status> {
+        let uid = Self::authenticate(&req)?;
+        let req = req.into_inner();
+        if req.args.is_empty() {
+            return Err(Status::invalid_argument("'args' must be a non-empty array"));
+        }
+        match Process::spawn(&*req.args[0], req.args[1..].iter().map(AsRef::as_ref)) {
+            Ok(p) => {
+                let pid = self.store.insert(&uid, p);
+                Ok(Response::new(ExecResponse {
+                    id: Some(pid.into()),
+                }))
+            }
+            Err(e) => Err(Status::unknown(format!("{}", e))),
+        }
     }
 
     async fn get_logs(
         &self,
-        _req: Request<LogsRequest>,
+        req: Request<LogsRequest>,
     ) -> Result<Response<Self::GetLogsStream>, Status> {
-        Err(Status::unimplemented(""))
+        let uid = Self::authenticate(&req)?;
+        let req = req.into_inner();
+        let pid = req
+            .id
+            .ok_or_else(|| Status::invalid_argument("Process ID not given"))?;
+        let process = self.get_process(pid, &uid)?;
+        // TODO: buffer multiple lines
+        let stream = process.logs().map(|b| Ok(LogsResponse { lines: vec![b] }));
+        Ok(Response::new(Box::pin(stream)))
     }
 
     async fn get_status(
         &self,
-        _req: Request<StatusRequest>,
+        req: Request<StatusRequest>,
     ) -> Result<Response<StatusResponse>, Status> {
-        Err(Status::unimplemented(""))
+        let uid = Self::authenticate(&req)?;
+        let req = req.into_inner();
+        let pid = req
+            .id
+            .ok_or_else(|| Status::invalid_argument("Process ID not given"))?;
+        let process = self.get_process(pid, &uid)?;
+        Ok(Response::new(StatusResponse {
+            exit_status: process.status().await.map(std_status_to_paas_status),
+        }))
     }
 
-    async fn stop(&self, _req: Request<StopRequest>) -> Result<Response<StopResponse>, Status> {
-        Err(Status::unimplemented(""))
+    async fn stop(&self, req: Request<StopRequest>) -> Result<Response<StopResponse>, Status> {
+        let uid = Self::authenticate(&req)?;
+        let req = req.into_inner();
+        let pid = req
+            .id
+            .ok_or_else(|| Status::invalid_argument("Process ID not given"))?;
+        let process = self.get_process(pid, &uid)?;
+        match process.stop().await {
+            Ok(_) => Ok(Response::new(StopResponse {})),
+            // TODO: aborted is a questionable status here
```

Comment:
I'd lean on having a notify so that this await is delayed until the ongoing stop operation is completed.

Reply:
Could you elaborate on this? Concurrent stops returning an error is per design. https://github.com/justinas/paas/blob/5dccc74379aa2ec7d3ab85e833b288d86b0af652/DESIGN.md#stop

Reply:
Oh right, I misread your comment in the earlier PR.